Refactor

average-gary opened the pull request:
Refactor hash functions and signing process in fullagg.py; improve clarity with comments and structure, and introduce KeyGen function for key pair generation.

I think the biggest change would be the H_sig method. Based on the document, all the L values preceded R, X_i, and m_i in the concatenation.

average-gary commented:
Add a README section.

I can clean up the commits/comments if desired. Thanks for writing this!

fjahr commented on fullagg.py:
```diff
@@ -242,38 +250,61 @@ def Ver(L: SignersList, sig: Signature) -> bool:
     m1, m2, m3 = b"jonas", b"tim", b"yannick"
 
     # DahLIAS round
+    # each signer runs (out_i, st_i) ← Sign() and sends out_i to the
```
I wouldn't put this here. This describes only part of what is happening in a full DahLIAS round and it is documented in other places. There could be more comments within DahLIAS but I have tried to keep the documentation mostly in the docstrings of the step functions. I find this the cleanest approach because otherwise if the same event (like sending out_i for example) is described twice in two different places a reader could find this confusing since it might be these are two different instances of the event that both have to happen rather than duplicate documentation. We document the same thing in different places in Bitcoin Core sometimes because you might miss stuff otherwise but this file is so small it should be easy to find the docs so I think it's better to document everything only once.

average-gary replied:
242a40af71224dfec8308e919fcff2c16fa2784a

fjahr commented on fullagg.py:
```diff
@@ -242,38 +250,61 @@ def Ver(L: SignersList, sig: Signature) -> bool:
+    # coordinator (note that pk_i, out_i, and m_i can be sent separately
+    # or all together to the coordinator);
     signers = [(sk1, pk1, m1), (sk2, pk2, m2), (sk3, pk3, m3)]
+
```
Hm, below in the tweaked signers case there is no empty line above the DahLIAS call. It would be ok for me to insert a line but then it should be consistent.

average-gary replied:
Reverted in 94a72343a12a0acc27c0234a1a70956c2a54336f

fjahr commented on fullagg.py:
```diff
@@ -242,38 +250,61 @@ def Ver(L: SignersList, sig: Signature) -> bool:
+    # to the coordinator;
+    (sk1, pk1), (sk2, pk2), (sk3, pk3) = KeyGen(), KeyGen(), KeyGen()
+
+    # each signer sends the message m_i it wants to sign to the coordinator;
```
(same as above) The sending only happens in the Coord step so I wouldn't put those comments here. We could say these are just dummy key pairs and messages for testing though.

fjahr commented on fullagg.py:
```diff
@@ -242,38 +250,61 @@ def Ver(L: SignersList, sig: Signature) -> bool:
 
     # Second coordinator round
-    sig = Coord2(R, s_list)
+    # Coord'
```
I find this redundant because in the docstring for the Coord2 function I already mention that this is called Coord' in the paper. Same for the other steps above, especially if the paper name isn't directly next to the function name.

average-gary replied:
Good call. Removing.

average-gary replied:
bcc7b80f0fa7bd8d1f43f86825c227e757ef5ad3

fjahr commented on fullagg.py:
```diff
@@ -242,38 +250,61 @@ def Ver(L: SignersList, sig: Signature) -> bool:
-    s_list = []
+    # Sign'
+    # Sign′ must be called at most once per signer state st_i
+    out_i_list = []
```
I don't understand this name change. out_i is defined as (R1_i, R2_i) but this is just a list of s values.

average-gary replied:
![Screenshot 2025-06-16 at 11 49 54](https://github.com/user-attachments/assets/fdf60c9e-ae08-438d-bc48-ecf91a4ec856)
I was basing it off the notation of the Sign' in the PDF as I followed the code with the specification. 
Makes sense either way so happy to revert.

fjahr replied:
hm, ok, then please just call it `out_list`, the `_i` in my notation means that it's a value associated to a specific `i`/signer and in a list it's implied that the values are all there so the `i` isn't needed in the name.

fjahr commented on fullagg.py:
```diff
@@ -196,7 +202,9 @@ def Coord(signer_inputs: List[Tuple[PublicKey, Message, SignerOutput]]) -> Tuple
     R = R1 + (b * R2)
 
-    return (ctx, R)
+    st = R
```
I found this a bit silly and I didn't handle it consistently but if we do this you might as well add a `SignerState` which is just a PublicNonce :)

average-gary replied:
There is already a `SignerState`, but given:

> Then, the coordinator stores state (st) and sends ctx to all signers.

I'll add `CommonNonceState = PublicNonce`?

average-gary replied:
fe7b7a06aaca7607902d1ac6d07b1f44b5b5d197

fjahr replied:
Actually I think I meant `CoordinatorState` here.

fjahr commented on fullagg.py:
```diff
@@ -86,7 +86,8 @@ def _rand_scalar() -> Scalar:
 
 ### Hash functions
-
+# a tagged hash of R1 || R2 || X_i || m_i || R2_i for all signers
```
You can add this to the docstring, same below in `hash_sig`

average-gary replied:
e36bc800bc05222d5204a8d33840a7986bcca0df

fjahr commented on fullagg.py:
```diff
@@ -242,38 +250,61 @@ def Ver(L: SignersList, sig: Signature) -> bool:
-    sk1, sk2, sk3 = _rand_scalar(), _rand_scalar(), _rand_scalar()
-    pk1, pk2, pk3 = sk1 * G, sk2 * G, sk3 * G
+    #  KeyGen
+    # each signer generates a key pair (sk_i, pk_i) ← KeyGen() and sends pk_i 
```
I would move the signing part as a docstring into the KeyGen function and leave the sending part to be documented within the DahLIAS function since that is clearer where in the sequence this happens.

average-gary replied:
I believe I address this in eeb699b4e413c3a0ebc74442c87225842236d075

fjahr reviewed (commented):
Very cool, thank you! I think the readme is definitely helpful and the changes in the code make sense, some of the renamings and additional comments are unnecessary though, I think. As I mention, I would like to avoid having the same events documented in multiple places, please see my comments and let me know what you think.

average-gary commented:
> Very cool, thank you! I think the readme is definitely helpful and the changes in the code make sense, some of the renamings and additional comments are unnecessary though, I think. As I mention, I would like to avoid having the same events documented in multiple places, please see my comments and let me know what you think.

Sounds good! I err'd for more verbose on first pass, hence the Draft status. I'll clean this up soon and ping you here when complete.

fjahr commented:
> Addressed all comments and tried to list the commit hash for each. Happy to rebase/squash but wanted narrowly scoped commits for each comment as appropriate.

I don't think you have actually pushed the changes here. Maybe you pushed it to a different branch? GitHub seems to be aware of some of the commit hashes you reference but not all. But generally with these kinds of changes there is no need to make separate fixup commits, I would rather like you to squash the changes into the one commit and mark my comments either as resolved or comment if you did something differently. This should also be less work for you.

average-gary commented:
> > Addressed all comments and tried to list the commit hash for each. Happy to rebase/squash but wanted narrowly scoped commits for each comment as appropriate.
> 
> I don't think you have actually pushed the changes here. Maybe you pushed it to a different branch? GitHub seems to be aware of some of the commit hashes you reference but not all. But generally with these kinds of changes there is no need to make separate fixup commits, I would rather like you to squash the changes into the one commit and mark my comments either as resolved or comment if you did something differently. This should also be less work for you.

Apologies, I pushed to a different branch than I used for this PR. I just pushed to the correct branch. Will circle back to your other comments soon.

edit: I'll also squash to one commit

fjahr commented:
ACK

Thanks a lot for contributing these improvements and sticking with it despite my many review comments!
